batchlib/analysis/background_extraction.py

```diff
@@ -120,53 +122,76 @@ def get_bg_stats(self, bg_values):
         assert n_cols == table.shape[1]
         return column_names, table
 
-    def write_stat_dict_to_talbe(self, stat_dict, first_column_name, table_key):
-        column_names, table = self.stat_dict_to_table(stat_dict, first_column_name)
+    def write_stat_dict_to_table(self, stat_dict, bg_fractions, first_column_name, table_key):
+        column_names, table = self.stat_dict_to_table(stat_dict, bg_fractions, first_column_name)
         with open_file(self.table_out_path, 'a') as f:
             self.write_table(f, table_key, column_names, table)
 
     def run(self, input_files, out_files):
-        # first, get plate wide and per-well background statistics
+        # first, get per image background statistics
         logger.info(f'{self.name}: computing background statistics')
-        bg_dict = {file: self.get_bg_segment(file, device='cpu') for file in input_files}
-        bg_per_image_stats = {in_file_to_image_name(file): self.get_bg_stats(bg_segments)
-                              for file, bg_segments in bg_dict.items()}
+
+        bg_dict = {file_name: self.get_bg_segment(file_name, device='cpu') for file_name in input_files}
+
+        # separate the bg dict into actual bg dict and background fractions
+        bg_fractions_per_image = {in_file_to_image_name(k): v[1] for k, v in bg_dict.items()}
+        bg_dict = {k: v[0] for k, v in bg_dict.items()}
+
+        bg_per_image_stats = {in_file_to_image_name(file_name): self.get_bg_stats(bg_segments)
+                              for file_name, bg_segments in bg_dict.items()}
 
         # ignore image outliers in per_well and per_plate backgrounds
         outlier_dict = _load_image_outliers(self.name, self.table_out_path, self.image_outlier_table, input_files)
+
         bg_per_well_dict = defaultdict(list)
+        bg_fractions_per_well = defaultdict(list)
+
         wells = set()
-        for file, bg_segment in bg_dict.items():
-            well = image_name_to_well_name(os.path.basename(file))
+        for file_name, bg_segment in bg_dict.items():
+            image_name = in_file_to_image_name(file_name)
+            well = image_name_to_well_name(image_name)
             wells.add(well)
-            image_name = os.path.splitext(os.path.split(file)[1])[0]
+
             if outlier_dict.get(image_name, (-1, None))[0] == 1:
                 logger.info(f'{self.name}: skipping outlier image in bg calculation')
                 continue
+
             bg_per_well_dict[well].append(bg_segment)
+            bg_fractions_per_well[well].append(bg_fractions_per_image[image_name])
+
         bg_per_well_dict = {well: torch.cat(bg_segments, dim=1) for well, bg_segments in bg_per_well_dict.items()}
         bg_per_well_stats = {well: self.get_bg_stats(bg_per_well_dict.get(well, None)) for well in wells}
+        bg_fractions_per_well = {well: np.median(bg_fractions_per_well[well]) if well in bg_fractions_per_well
+                                 else None for well in wells}
```

Comment:
If think this should be the mean (at least as long as all images have the same size, that should be the right way to compute the bg fraction over all images per well).

Reply:
Yes, makes sense.